src/Autofac/Core/Resolving/InstanceLookup.cs

```diff
@@ -86,10 +86,10 @@ public object Execute()
 
             var resolveParameters = Parameters as Parameter[] ?? Parameters.ToArray();
 
-            if (!_activationScope.TryGetSharedInstance(ComponentRegistration.Id, out var instance))
+            if (!_activationScope.TryGetSharedInstance(ComponentRegistration.Id, _decoratorTargetComponent?.Id, out var instance))
```

Comment:
Making the second parameter nullable made this change much more simple, otherwise it'd be a nasty collection of nested conditions and near copy/paste logic.